public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -518,27 +518,35 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
       .map(label => `${label}="${row.labels[label]}"`)
       .join(',');
 
-    const contextTimeBuffer = 2 * 60 * 60 * 1000 * 1e6; // 2h buffer
-    const timeEpochNs = row.timeEpochMs * 1e6;
+    const contextTimeBuffer = 2 * 60 * 60 * 1000; // 2h buffer
     const commonTargetOptions = {
       limit,
       query: `{${query}}`,
       expr: `{${query}}`,
       direction,
     };
 
+    const fieldCache = new FieldCache(row.dataFrame);
+    const nsField = fieldCache.getFieldByName('tsNs')!;
+    const nsTimestamp = nsField.values.get(row.rowIndex);
+
     if (direction === 'BACKWARD') {
       return {
         ...commonTargetOptions,
-        start: timeEpochNs - contextTimeBuffer,
-        end: timeEpochNs, // using RFC3339Nano format to avoid precision loss
+        // convert to ns, we loose some precision here but it is not that important at the far points of the context
+        start: row.timeEpochMs - contextTimeBuffer + '000000',
```

Comment:
`* 1e6` would work, but as other operations with ns precision integers may not work as it is over SAFE_INTEGER limit, it seems clearer to just append this as a string.